cpp/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -2229,6 +2229,11 @@ Status ConvertChunkedArrayToPandas(const PandasOptions& options,
         checked_cast<const DictionaryType&>(*arr->type()).value_type();
     RETURN_NOT_OK(DecodeDictionaries(options.pool, dense_type, &arr));
     DCHECK_NE(arr->type()->id(), Type::DICTIONARY);
+
+    // The original Python DictionaryArray won't own the memory anymore
+    // as we actually built a new array when we decoded the DictionaryArray
+    // thus let the final resulting numpy array own the memory through a Capsule
+    py_ref = nullptr;
```

Comment:
@pitrou would you be able to confirm that this is the right fix?

You can see the problem it deals with by checking out this branch, commenting out that line and running the `test_dictionary_to_numpy` test. You will see that in such case we produce corrupted data as the subsequent `MakeNumPyView` used the original `DictionaryArray` as the "base" of the new `numpy` array and the newly built `Array` resulting by dictionary decoding (who really owns the memory for the numpy array) probably goes away immediately at the end of this method. With this change instead a capsule is made and the returned numpy array becomes the owner of that capsule.

Does that sound like the right solution?

Reply:
This does seem right indeed. Kudos for figuring this out :-)